### PR TITLE
Support multiple-domains in MINIO_DOMAIN

### LINKED ISSUE #7173: Which part of the code handles 405 Method Not Allowed
Hi.
I am trying to change the server code in a way that server could handle multiple domains for one instance (an array of `globalDomainName`)
For that I handled routers and methods in `cmd/api-router.go`
but I cant find which part of the code handles endpoint which when I enter `export MINIO_DOMAIN=mydomain.dev` handles it and allows `mybucket.mydomain.dev/` but it returns ` 405 Method Not Allowed.` for lets say `mybucket.myseconddomain.dev/`.

Is it possible for you to help me on this? or suggesting another approach which hadnt come to my mind?

tnx.

ps: I traced MINIO_DOMAIN and reached to globalDomainName variable and followed that but couldnt locate what I was looking for.

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Support multiple-domains in MINIO_DOMAIN

## Motivation and Context
Fixes #7173

## Regression
No

## How Has This Been Tested?
Just set two different domains as a comma-separated list
```
~  MINIO_DOMAIN=localhost,backspace minio server ~/test
```
```
~ cat /etc/hosts
127.0.0.1	localhost
127.0.0.1       backspace
127.0.0.1       testbucket.backspace
```

```
mc: <DEBUG> GET /?delimiter=%2F&max-keys=1000&prefix= HTTP/1.1
Host: testbucket.localhost:9000
User-Agent: Minio (linux; amd64) minio-go/v6.0.17 mc/2019-02-22T08:25:28Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20190222/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20190222T093040Z
Accept-Encoding: gzip
```
```
mc: <DEBUG> GET /?delimiter=%2F&max-keys=1000&prefix= HTTP/1.1
Host: testbucket.backspace:9000
User-Agent: Minio (linux; amd64) minio-go/v6.0.17 mc/2019-02-22T08:25:28Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20190222/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20190222T093354Z
Accept-Encoding: gzip
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#849e8a31-4944-454c-a012-294eb3771c36) |
<a name="849e8a31-4944-454c-a012-294eb3771c36"></a>
### 7274-74891c5/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:31701
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3d7f1d688854:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 42 seconds
(4/13) Running aws-sdk-ruby tests ... done in 2 seconds
(5/13) Running awscli tests ... done in 1 minutes and 42 seconds
(6/13) Running mc tests ... done in 16 seconds
(7/13) Running minio-dotnet tests ... done in 54 seconds
(8/13) Running minio-go tests ... done in 19 seconds
(9/13) Running minio-java tests ... FAILED in 24 seconds
{
  "name": "minio-java",
  "function": "putObject(String bucketName, String objectName, InputStream body, long size, String contentType)",
  "args": "contentType: application/javascript",
  "duration": 19801,
  "status": "FAIL",
  "error": "java.net.SocketException: Connection reset >>> [java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:115), java.net.SocketOutputStream.write(SocketOutputStream.java:155), okio.Okio$1.write(Okio.java:78), okio.AsyncTimeout$1.write(AsyncTimeout.java:179), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:171), okio.RealBufferedSink.write(RealBufferedSink.java:41), okhttp3.internal.http1.Http1Codec$FixedLengthSink.write(Http1Codec.java:286), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:171), okio.RealBufferedSink.write(RealBufferedSink.java:109), io.minio.HttpRequestBody.writeTo(HttpRequestBody.java:74), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:62), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:185), okhttp3.RealCall.execute(RealCall.java:69), io.minio.MinioClient.executeReq(MinioClient.java:1097), io.minio.MinioClient.execute(MinioClient.java:1056), io.minio.MinioClient.executePut(MinioClient.java:1399), io.minio.MinioClient.executePut(MinioClient.java:1421), io.minio.MinioClient.putObject(MinioClient.java:3994), io.minio.MinioClient.putObject(MinioClient.java:4086), io.minio.MinioClient.putObject(MinioClient.java:3960), FunctionalTest.testPutObjectUnknownStreamSize(FunctionalTest.java:556), FunctionalTest.putObject_test8(FunctionalTest.java:582), FunctionalTest.runTests(FunctionalTest.java:2998), FunctionalTest.main(FunctionalTest.java:3154)]"
}

Executed 8 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

**Review by nitisht — CHANGES_REQUESTED** [review]
we need to update the docs as well https://github.com/minio/minio/tree/master/docs/config#domain

also PTAL at the Travis failure

File: cmd/common-main.go [hunk 1]
```diff
@@ -262,10 +262,14 @@ func handleCommonEnvVars() {
 		logger.FatalIf(err, "Unable to initialize etcd with %s", etcdEndpoints)
 	}
 
-	globalDomainName, globalIsEnvDomainName = os.LookupEnv("MINIO_DOMAIN")
-	if globalDomainName != "" {
-		if _, ok = dns2.IsDomainName(globalDomainName); !ok {
-			logger.Fatal(uiErrInvalidDomainValue(nil).Msg("Unknown value `%s`", globalDomainName), "Invalid MINIO_DOMAIN value in environment variable")
+	v, ok := os.LookupEnv("MINIO_DOMAIN")
```
Comment (poornas) [inline cmd/common-main.go:265]: this should be MINIO_DOMAINS for consistency across envs

**Review by poornas — CHANGES_REQUESTED** [review]

File: cmd/common-main.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/common-main.go:265]: There is a good reason it's not DOMAINS as it's not multiple domains simultaneously accessing Minio it is always a DOMAIN. The point is that they all are just DOMAIN for Minio at the request level. It is similar to `--encrypt-key` option we don't say it `--encrypt-keys` or `--encrypts` for `--encrypt` option.

An additional benefit is avoiding breakage in backward compatibility, most users won't need this feature anyways.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#07f57ccd-c6d4-44d3-90c9-707d5e1f02a1) |
<a name="07f57ccd-c6d4-44d3-90c9-707d5e1f02a1"></a>
### 7274-090c81d/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:31552
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b7e17c53da71:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... FAILED in 2 minutes and 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-skuw9h0vbid2ld",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 120330,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.60:31552/aws-sdk-go-test-skuw9h0vbid2ld: dial tcp 72.28.97.60:31552: i/o timeout",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 13 tests successfully.
```

File: cmd/common-main.go (hunk 1 above)
Comment (poornas) [reply cmd/common-main.go:265]: that's a good argument

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#1ffa642e-31af-4b44-83bb-c7981b6900e1) |
| mint-large-bucket.sh | :x: [more...](#96966b45-cb30-417c-a903-37aa1081c8d6) |
| mint-compression-dist-xl.sh | :x: [more...](#92490dd5-b607-45ba-a646-777de80e106b) |
| mint-dist-xl.sh | :x: [more...](#5de1ac37-7e8a-40cd-90a6-b1bcf879268c) |
<a name="5de1ac37-7e8a-40cd-90a6-b1bcf879268c"></a>
### 7274-72f3c4f/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:30992
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6f5369c00093:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... FAILED in 6 minutes and 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-a2mxwqccg95bpl",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 360024,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 1585CD773966159D, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 13 tests successfully.
```
<a name="92490dd5-b607-45ba-a646-777de80e106b"></a>
### 7274-72f3c4f/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:30535
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c021bff1d211:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... FAILED in 2 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-tl1h2dcr1vs15s",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 1364,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.60:30535/aws-sdk-go-test-tl1h2dcr1vs15s: dial tcp 72.28.97.60:30535: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 13 tests successfully.
```
<a name="96966b45-cb30-417c-a903-37aa1081c8d6"></a>
### 7274-72f3c4f/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:32146
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp eaf337109acd:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-b3cjkpltkmew1v",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 309,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.51:32146/aws-sdk-go-test-b3cjkpltkmew1v: dial tcp 72.28.97.51:32146: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 13 tests successfully.
```
<a name="1ffa642e-31af-4b44-83bb-c7981b6900e1"></a>
### 7274-72f3c4f/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:31624
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 2c4a3efb58ac:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-92wu5wj6pbkiwr",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 351,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.60:31624/aws-sdk-go-test-92wu5wj6pbkiwr: dial tcp 72.28.97.60:31624: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 13 tests successfully.
```

**Review by poornas — APPROVED** [review]
LGTM

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#25116236-754e-46a0-b074-7d62acb46806) |
<a name="25116236-754e-46a0-b074-7d62acb46806"></a>
### 7274-410abda/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:32561
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp dad072ac2890:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 44 seconds
(4/13) Running aws-sdk-ruby tests ... FAILED in 1 minutes and 4 seconds
{
  "name": "aws-sdk-ruby",
  "function": "presignedPut(bucket_name,file_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-38960cb1ee16",
    "file_name": "datafile-1-MB"
  },
  "duration": 60029.27,
  "error": "execution expired",
  "status": "FAIL"
}

Executed 3 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#330685f0-6940-43d3-bd69-b83d2a744b24) |
<a name="330685f0-6940-43d3-bd69-b83d2a744b24"></a>
### 7274-410abda/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:30364
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 5ec4bad5a857:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... FAILED in 2 minutes and 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-m4ellfkclxbrxn",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 120312,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.60:30364/aws-sdk-go-test-m4ellfkclxbrxn: dial tcp 72.28.97.60:30364: i/o timeout",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 13 tests successfully.
```

**Review by nitisht — APPROVED** [review]

**Pull request merged**